### ISSUE TITLE
Standby node fails to start after restart due to existend tablespace

### ISSUE THREAD

**Issue by oPhilMehr** [bitnami/containers#21548]
### Name and Version

bitnami/postgresql-repmgr:14

### What steps will reproduce the bug?

Setup: Two nodes as a docker swarm (yes, 2 nodes, this is a PoC) and bitnami/postgresql-repmgr:14, one running on each node (with placement constraints)
Persistent volume mount `/db:/bitnami/postgresql`

Content of directory `/bitnami/postgresql` (or `/db`):
```console
$ ls -l /db
total 4
drwxr-xr-x.  2 1001 root   48 Jan 24 13:15 conf
drwx------. 19 1001 root 4096 Jan 24 13:36 data
drwxr-xr-x.  2 1001 root    6 Jan 23 17:13 lock
drwx------.  3 1001 root   17 Jan 24 13:36 tablespace
```

View from pgAdmin:
![grafik](https://user-images.githubusercontent.com/82316355/214319261-bf91ab5f-ba44-4119-b2d4-1a79f37de2fb.png)

Situation: Primary is running and standby too. Now we restart the standby node, see "what do you see instead" for logs. In short: Standby tries to execute a `pg_basebackup` (from primary?) and fails because under folder `/bitnami/postgresql/tablespace` data is existent.
So Docker Swarm restarts the container. This is an infinite loop.

You can fix this by removing the "tablespace" folder or its content. But that does not make sense, think of a 200 GB database/tablespace and you have to delete/copy this data on each restart or failover situation.

Where is my mistake? Wrong location for the tablespace folder?
Or is it a bug?

Many thanks for your time and feedback!

Best regards,
Philip

### What is the expected behavior?

The Standby node should come up without any manual tasks.

### What do you see instead?

Output on the standby node:
```log
postgresql-repmgr 16:01:01.65 
postgresql-repmgr 16:01:01.65 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 16:01:01.65 Subscribe to project updates by watching https://github.com/bitnami/containers
postgresql-repmgr 16:01:01.66 Submit issues and feature requests at https://github.com/bitnami/containers/issues
postgresql-repmgr 16:01:01.66 
postgresql-repmgr 16:01:01.66 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 16:01:01.68 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 16:01:01.70 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 16:01:01.70 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 16:01:01.71 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 16:01:01.71 DEBUG ==> Checking node 'node-01:5432'...
postgresql-repmgr 16:01:01.76 DEBUG ==> Pretending primary role node - 'node-01:5432'
postgresql-repmgr 16:01:01.76 DEBUG ==> Pretending primary set to 'node-01:5432'!
postgresql-repmgr 16:01:01.76 DEBUG ==> Checking node 'node-02:5432'...
psql: error: could not translate host name "node-02" to address: Temporary failure in name resolution
postgresql-repmgr 16:01:01.80 DEBUG ==> Skipping: failed to get primary from the node 'node-02:5432'!
postgresql-repmgr 16:01:01.80 INFO  ==> Auto-detected primary node: 'node-01:5432'
postgresql-repmgr 16:01:01.80 INFO  ==> This node was acting as a primary before restart!
postgresql-repmgr 16:01:01.80 INFO  ==> Current master is 'node-01:5432'. Cloning/rewinding it and acting as a standby node...
postgresql-repmgr 16:01:01.81 DEBUG ==> Primary node: 'node-01:5432'
postgresql-repmgr 16:01:01.81 INFO  ==> Node configured as standby
postgresql-repmgr 16:01:01.82 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 16:01:01.82 DEBUG ==> Injecting a new postgresql.conf file...
postgresql-repmgr 16:01:01.82 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 16:01:01.93 DEBUG ==> Injecting a new pg_hba.conf file...
postgresql-repmgr 16:01:01.93 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 16:01:01.94 DEBUG ==> Node ID: '2', Rol: 'standby', Primary Node: 'node-01:5432'
postgresql-repmgr 16:01:01.95 INFO  ==> Initializing Repmgr...
postgresql-repmgr 16:01:01.95 INFO  ==> Waiting for primary node...
postgresql-repmgr 16:01:01.96 DEBUG ==> Wait for schema repmgr.repmgr on 'node-01:5432', will try 6 times with 10 delay seconds (TIMEOUT=60)
postgresql-repmgr 16:01:01.98 DEBUG ==> Schema repmgr.repmgr exists!
postgresql-repmgr 16:01:01.98 INFO  ==> Rejoining node...
postgresql-repmgr 16:01:01.99 INFO  ==> Cloning data from primary node...
WARNING: following problems with command line parameters detected:
  -D/--pgdata will be ignored if a repmgr configuration file is provided
NOTICE: destination directory "/bitnami/postgresql/data" provided
NOTICE: checking for available walsenders on the source node (2 required)
NOTICE: checking replication connections can be made to the source server (2 required)
WARNING: data checksums are not enabled and "wal_log_hints" is "off"
DETAIL: pg_rewind requires "wal_log_hints" to be enabled
WARNING: directory "/bitnami/postgresql/data" exists but is not empty
NOTICE: -F/--force provided - deleting existing data directory "/bitnami/postgresql/data"
NOTICE: starting backup (using pg_basebackup)...
pg_basebackup: error: directory "/bitnami/postgresql/tablespace/DER" exists but is not empty
pg_basebackup: removing contents of data directory "/bitnami/postgresql/data"
NOTICE: replication slot "repmgr_slot_2" deleted on node -1
ERROR: unable to take a base backup of the source server
HINT: data directory ("/bitnami/postgresql/data") may need to be cleaned up manually
#### new docker container
postgresql-repmgr 16:01:08.16 
postgresql-repmgr 16:01:08.16 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 16:01:08.17 Subscribe to project updates by watching https://github.com/bitnami/containers
postgresql-repmgr 16:01:08.17 Submit issues and feature requests at https://github.com/bitnami/containers/issues
postgresql-repmgr 16:01:08.17 
```

and output will be the same (with new docker container) until you remove the tablespace "DER".

On the primary node there are also more then less interesting log entries:
```log
2023-01-24 16:01:02.244 CET [394] LOG:  could not send data to client: Broken pipe
2023-01-24 16:01:02.244 CET [394] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:02.244 CET [394] ERROR:  base backup could not send data, aborting backup
2023-01-24 16:01:02.244 CET [394] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:02.245 CET [394] FATAL:  connection to client lost
2023-01-24 16:01:08.665 CET [407] LOG:  could not send data to client: Connection reset by peer
2023-01-24 16:01:08.665 CET [407] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:08.665 CET [407] ERROR:  base backup could not send data, aborting backup
2023-01-24 16:01:08.665 CET [407] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:08.665 CET [407] LOG:  could not send data to client: Broken pipe
2023-01-24 16:01:08.665 CET [407] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:08.665 CET [407] FATAL:  connection to client lost
2023-01-24 16:01:15.046 CET [420] LOG:  could not send data to client: Connection reset by peer
2023-01-24 16:01:15.046 CET [420] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:15.046 CET [420] ERROR:  base backup could not send data, aborting backup
2023-01-24 16:01:15.046 CET [420] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:15.046 CET [420] LOG:  could not send data to client: Broken pipe
2023-01-24 16:01:15.046 CET [420] STATEMENT:  BASE_BACKUP LABEL 'repmgr base backup' PROGRESS  FAST NOWAIT    MANIFEST 'yes' 
2023-01-24 16:01:15.046 CET [420] FATAL:  connection to client lost
```

Log entries of primary are caused by standby trying to connect, I think the "connection lost" and "broken pipe" is because of the standby node failing and restarting.

### Additional information

Docker compose file:
```yml
  node-01:
    image: bitnami/postgresql-repmgr:14
    volumes:
    - /db:/bitnami/postgresql
    environment:
    - POSTGRESQL_PASSWORD=secret
    - REPMGR_PASSWORD=otherSecret
    - REPMGR_PRIMARY_HOST=node-01
    - REPMGR_PARTNER_NODES=node-01,node-02
    - BITNAMI_DEBUG=true
    - REPMGR_NODE_NAME=node-01
    - REPMGR_NODE_ID=1
    - REPMGR_NODE_NETWORK_NAME=node-01
    deploy:
      placement:
        constraints:
        - node.hostname == MACHINE01
      restart_policy:
        condition: on-failure

  node-02:
    image: bitnami/postgresql-repmgr:14
    volumes:
    - /db:/bitnami/postgresql
    environment:
    - POSTGRESQL_PASSWORD=secret
    - REPMGR_PASSWORD=otherSecret
    - REPMGR_PRIMARY_HOST=node-01
    - REPMGR_PARTNER_NODES=node-01,node-02
    - BITNAMI_DEBUG=true
    - REPMGR_NODE_NAME=node-02
    - REPMGR_NODE_ID=2
    - REPMGR_NODE_NETWORK_NAME=node-02
    deploy:
      placement:
        constraints:
        - node.hostname == MACHINE02
      restart_policy:
        condition: on-failure
```

I also stopped the primary node, failover to standby was successful. Afterward, I restarted the standby (primary now) again and it's just working fine.
Then I started the stopped primary node and the same error occurs, same log file entries as the other way around (restarting standby while primary is running). And here is the same thing: Removing the content of the "tablespace" folder will fix the infinite restart loop.

**Comment by oPhilMehr** [comment]
Hi all,
I did a few more tests and got the following result:

I set up an identical setup with postgres and repmgr, expect that I didn't use an "external" table space like DER in the scenario above. So Postgres uses the default tablespace to store newly created databases.

After creating a database with some random data I restarted the second node and started it again. It will also perform a pg_basebackup. But it's working very fine. Standby won't fail, instead it reconnects successful to the "cluster".

Here logs of standby node
```log
postgresql-repmgr 18:17:40.07 
postgresql-repmgr 18:17:40.07 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 18:17:40.07 Subscribe to project updates by watching https://github.com/bitnami/containers
postgresql-repmgr 18:17:40.07 Submit issues and feature requests at https://github.com/bitnami/containers/issues
postgresql-repmgr 18:17:40.08 
postgresql-repmgr 18:17:40.13 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 18:17:40.16 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 18:17:40.18 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 18:17:40.19 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 18:17:40.19 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 18:17:40.20 DEBUG ==> Checking node 'node-01:5432'...
postgresql-repmgr 18:17:40.39 DEBUG ==> Pretending primary role node - 'node-01:5432'
postgresql-repmgr 18:17:40.39 DEBUG ==> Pretending primary set to 'node-01:5432'!
postgresql-repmgr 18:17:40.40 DEBUG ==> Checking node 'node-02:5432'...
psql: error: connection to server at "node-02" (10.0.2.232), port 5432 failed: Connection refused
   Is the server running on that host and accepting TCP/IP connections?
postgresql-repmgr 18:17:40.41 DEBUG ==> Skipping: failed to get primary from the node 'node-02:5432'!
postgresql-repmgr 18:17:40.41 INFO  ==> Auto-detected primary node: 'node-01:5432'
postgresql-repmgr 18:17:40.42 DEBUG ==> Primary node: 'node-01:5432'
postgresql-repmgr 18:17:40.43 INFO  ==> Node configured as standby
postgresql-repmgr 18:17:40.44 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 18:17:40.45 DEBUG ==> Injecting a new postgresql.conf file...
postgresql-repmgr 18:17:40.46 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 18:17:40.63 DEBUG ==> Injecting a new pg_hba.conf file...
postgresql-repmgr 18:17:40.64 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 18:17:40.65 DEBUG ==> Node ID: '2', Rol: 'standby', Primary Node: 'node-01:5432'
postgresql-repmgr 18:17:40.65 INFO  ==> Initializing Repmgr...
postgresql-repmgr 18:17:40.66 INFO  ==> Waiting for primary node...
postgresql-repmgr 18:17:40.66 DEBUG ==> Wait for schema repmgr.repmgr on 'node-01:5432', will try 6 times with 10 delay seconds (TIMEOUT=60)
postgresql-repmgr 18:17:40.70 DEBUG ==> Schema repmgr.repmgr exists!
postgresql-repmgr 18:17:40.70 INFO  ==> Rejoining node...
postgresql-repmgr 18:17:40.70 INFO  ==> Cloning data from primary node...
WARNING: following problems with command line parameters detected:
  -D/--pgdata will be ignored if a repmgr configuration file is provided
NOTICE: destination directory "/bitnami/postgresql/data" provided
NOTICE: checking for available walsenders on the source node (2 required)
NOTICE: checking replication connections can be made to the source server (2 required)
WARNING: data checksums are not enabled and "wal_log_hints" is "off"
DETAIL: pg_rewind requires "wal_log_hints" to be enabled
WARNING: directory "/bitnami/postgresql/data" exists but is not empty
NOTICE: -F/--force provided - deleting existing data directory "/bitnami/postgresql/data"
NOTICE: starting backup (using pg_basebackup)...
NOTICE: standby clone (using pg_basebackup) complete
NOTICE: you can now start your PostgreSQL server
HINT: for example: pg_ctl -D /bitnami/postgresql/data start
HINT: after starting the server, you need to re-register this standby with "repmgr standby register --force" to update the existing node record
[REPMGR EVENT] Node id: 2; Event type: standby_clone; Success [1|0]: 1; Time: 2023-01-25 18:17:42.4323+01;  Details: cloned from host "node-01", port 5432; backup method: pg_basebackup; --force: Y
Looking for the script: /opt/bitnami/repmgr/events/execs/standby_clone.sh
[REPMGR EVENT] no script '/opt/bitnami/repmgr/events/execs/standby_clone.sh' found. Skipping...
postgresql-repmgr 18:17:42.54 INFO  ==> Initializing PostgreSQL database...
postgresql-repmgr 18:17:42.56 DEBUG ==> Copying files from /bitnami/postgresql/conf to /opt/bitnami/postgresql/conf
postgresql-repmgr 18:17:42.57 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql-repmgr 18:17:42.57 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql-repmgr 18:17:42.58 DEBUG ==> Ensuring expected directories/files exist...
postgresql-repmgr 18:17:42.66 INFO  ==> Deploying PostgreSQL with persisted data...
postgresql-repmgr 18:17:42.69 INFO  ==> Configuring replication parameters
postgresql-repmgr 18:17:42.72 INFO  ==> Configuring fsync
postgresql-repmgr 18:17:42.74 INFO  ==> Setting up streaming replication slave...
postgresql-repmgr 18:17:42.82 INFO  ==> Starting PostgreSQL in background...
waiting for server to start....2023-01-25 18:17:43.158 CET [200] LOG:  pgaudit extension initialized
2023-01-25 18:17:43.165 CET [200] LOG:  redirecting log output to logging collector process
2023-01-25 18:17:43.165 CET [200] HINT:  Future log output will appear in directory "/opt/bitnami/postgresql/logs".
2023-01-25 18:17:43.165 CET [200] LOG:  starting PostgreSQL 14.6 on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
2023-01-25 18:17:43.166 CET [200] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-01-25 18:17:43.166 CET [200] LOG:  listening on IPv6 address "::", port 5432
2023-01-25 18:17:43.168 CET [200] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2023-01-25 18:17:43.171 CET [202] LOG:  database system was interrupted; last known up at 2023-01-25 18:17:41 CET
2023-01-25 18:17:43.408 CET [202] LOG:  entering standby mode
2023-01-25 18:17:43.410 CET [202] LOG:  redo starts at 0/9000028
2023-01-25 18:17:43.411 CET [202] LOG:  consistent recovery state reached at 0/9000100
2023-01-25 18:17:43.412 CET [200] LOG:  database system is ready to accept read-only connections
2023-01-25 18:17:43.430 CET [206] LOG:  started streaming WAL from primary at 0/A000000 on timeline 1
 done
server started
postgresql-repmgr 18:17:43.53 INFO  ==> Unregistering standby node...
NOTICE: unregistering node 2
[REPMGR EVENT] Node id: 2; Event type: standby_unregister; Success [1|0]: 1; Time: 2023-01-25 18:17:43.593547+01;  Details: 
Looking for the script: /opt/bitnami/repmgr/events/execs/standby_unregister.sh
[REPMGR EVENT] no script '/opt/bitnami/repmgr/events/execs/standby_unregister.sh' found. Skipping...
postgresql-repmgr 18:17:43.60 INFO  ==> Registering Standby node...
NOTICE: using provided configuration file "/opt/bitnami/repmgr/conf/repmgr.conf"
WARNING: --upstream-node-id not supplied, assuming upstream node is primary (node ID: 1)
[REPMGR EVENT] Node id: 2; Event type: standby_register; Success [1|0]: 1; Time: 2023-01-25 18:17:43.660096+01;  Details: standby registration succeeded; upstream node ID is 1 (-F/--force option was used)
Looking for the script: /opt/bitnami/repmgr/events/execs/standby_register.sh
[REPMGR EVENT] no script '/opt/bitnami/repmgr/events/execs/standby_register.sh' found. Skipping...
NOTICE: standby node "node-02" (ID: 2) successfully registered
postgresql-repmgr 18:17:43.67 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down....2023-01-25 18:17:43.679 CET [200] LOG:  received fast shutdown request
2023-01-25 18:17:43.681 CET [200] LOG:  aborting any active transactions
2023-01-25 18:17:43.681 CET [206] FATAL:  terminating walreceiver process due to administrator command
2023-01-25 18:17:43.682 CET [203] LOG:  shutting down
2023-01-25 18:17:43.687 CET [200] LOG:  database system is shut down
 done
server stopped
postgresql-repmgr 18:17:43.78 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **

postgresql-repmgr 18:17:43.83 INFO  ==> Starting PostgreSQL in background...
waiting for server to start....2023-01-25 18:17:43.856 CET [235] LOG:  pgaudit extension initialized
2023-01-25 18:17:43.862 CET [235] LOG:  redirecting log output to logging collector process
2023-01-25 18:17:43.862 CET [235] HINT:  Future log output will appear in directory "/opt/bitnami/postgresql/logs".
2023-01-25 18:17:43.862 CET [235] LOG:  starting PostgreSQL 14.6 on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
2023-01-25 18:17:43.863 CET [235] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-01-25 18:17:43.863 CET [235] LOG:  listening on IPv6 address "::", port 5432
2023-01-25 18:17:43.865 CET [235] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2023-01-25 18:17:43.870 CET [237] LOG:  database system was shut down in recovery at 2023-01-25 18:17:43 CET
2023-01-25 18:17:43.871 CET [237] LOG:  entering standby mode
2023-01-25 18:17:43.874 CET [237] LOG:  redo starts at 0/9000028
2023-01-25 18:17:43.875 CET [237] LOG:  consistent recovery state reached at 0/A000E78
2023-01-25 18:17:43.875 CET [237] LOG:  invalid record length at 0/A000E78: wanted 24, got 0
2023-01-25 18:17:43.875 CET [235] LOG:  database system is ready to accept read-only connections
2023-01-25 18:17:43.889 CET [241] LOG:  started streaming WAL from primary at 0/A000000 on timeline 1
 done
server started
postgresql-repmgr 18:17:43.95 INFO  ==> ** Starting repmgrd **
[2023-01-25 18:17:44] [NOTICE] repmgrd (repmgrd 5.3.3) starting up
INFO:  set_repmgrd_pid(): provided pidfile is /tmp/repmgrd.pid
[2023-01-25 18:17:44] [NOTICE] starting monitoring of node "node-02" (ID: 2)
```

The primary only logs "node2 disconnected" and "node2 reconnected after x seconds"

Thank you,

Best regards,
Philip

**Comment by carrodher** [comment]
In the end, in both cases, the initialization logic is running the same commands until reaching
```
INFO  ==> Cloning data from primary node...
```
But in the first case fails when trying to backup the `/bitnami/postgresql/tablespace/DER` directory:
```
WARNING: directory "/bitnami/postgresql/data" exists but is not empty
NOTICE: -F/--force provided - deleting existing data directory "/bitnami/postgresql/data"
NOTICE: starting backup (using pg_basebackup)...
pg_basebackup: error: directory "/bitnami/postgresql/tablespace/DER" exists but is not empty
pg_basebackup: removing contents of data directory "/bitnami/postgresql/data"
NOTICE: replication slot "repmgr_slot_2" deleted on node -1
ERROR: unable to take a base backup of the source server
HINT: data directory ("/bitnami/postgresql/data") may need to be cleaned up manually
```
while in the second case, the backup is done without any issues.

I'm not sure if there is something that can be modified in the container logic present at https://github.com/bitnami/containers/blob/main/bitnami/postgresql-repmgr/14/debian-11/rootfs/opt/bitnami/scripts/librepmgr.sh#L612 in order to remove any other directory causing this issue before starting the clone/backup. Would you like to give it a try and send a PR?

**Comment by oPhilMehr** [comment]
Hello carrodher,

Thank you for your investigation and feedback.

I already had a look to the part of the script and it's logic but no intention to implement anything. Thought it must be something else, wrong paths for tablespaces for example.
Does no one else have this problem? No one use tablespaces? Or don't you restart a node?
This issue is closest to the problem here: https://github.com/bitnami/charts/issues/8705

Seems to be a problem with postgres pg_basebackup. Here are some more results of my investigation from the web. Looks like the problem have been around since 2015, most of the threads where not complete or never answered:
https://groups.google.com/g/repmgr/c/MxutWFt9NkY
https://groups.google.com/g/repmgr/c/iCkfhjbWw9g
https://dba.stackexchange.com/questions/95204/postgres-pg-basebackup-reports-directory-exists-but-is-not-empty

It is very suspicious that pg_basebackup deletes the pgdata-directory of postgres where the default tablespace is located.
- When you create your tablespace inside the pgdata directory (which is not recommended), the basebackup will still fail even though the pgdata directory is deleted.
- And why won't a user defined tablespace (directory) be deleted..

The problem with implementing a logic that deletes the user defined tablespace is, that you never get the location of the tablespaces via script. This is one ore more values provided by a human and could be `/walla` or `/foo/bar`.
We could define a "default" location under which all tablespaces should be created. I will try this out and provide feedback soon.

Thank you,

Best regards,
Philip

**Comment by CeliaGMqrz** [comment]
Hi @oPhilMehr, 

Thanks for the detailed information. 

According to [upstream](https://www.postgresql.org/docs/current/manage-ag-tablespaces.html):

> The directory $PGDATA/pg_tblspc contains symbolic links that point to each of the non-built-in tablespaces defined in the cluster. Although not recommended, it is possible to adjust the tablespace layout by hand by redefining these links. Under no circumstances perform this operation while the server is running. Note that in PostgreSQL 9.1 and earlier you will also need to update the pg_tablespace catalog with the new locations. (If you do not, pg_dump will continue to output the old tablespace locations.)

I'm not sure, but I think this problem comes from this premise. Note that although the tablespace may be located in a directory outside the data directory, it is still part of the database cluster. If the tablespaces you create are located outside the data directory, `pg_basebackup` will not be able to save them in the backup directory. The reason is that this conflicts with the data directory structure because the link in `pg_tblspc` still points to the path provided.

You can try this to get more information:
* Check tablespace location `SELECT oid, spcname FROM pg_tablespace`
* You can test pg_basebackup to debug `pg_basebackup -U <user> -D /path/to/backup --xlog-method=stream --tablespace-mapping <tablespace_oid>=/path/to/tablespaceLocation`

We have not received any issue similar to this one, and it is not trivial to reproduce it, but if this issue receives more demand we could create a task to investigate it. I will leave the case open for more people in the community to contribute. If you have any updates or find a solution we will be happy to receive your feedback.

**Comment by diego2glez** [comment]
I need to reopen the issue as i'm stuck with the same problem here. Couldn't image postgresql-repmgr:15 manage replication of tablespaces created outside the /bitnami/postgresql/data default path?

Using the docker-compose provided in the repo I attached two new volumes in same path for both postgres backends.

pg-0 (primary):
- pg_0_oapi_data:/data/oapi/

pg-1 (standby):
- pg_1_oapi_data:/data/oapi/

I created the directory /data/oapi/ts_oapi 

After start the cluster I create a tablespace pointing to that directory, a database and add some data. Replication works as expected.

Then after restart the cluster in debug, the standby backend log shows the following messages in loop.
...
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | postgresql-repmgr 13:10:15.84 INFO  ==> Cloning data from primary node...
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | WARNING: following problems with command line parameters detected:
pg-test_pg-1.1.aqsq6efqjcvi@docker03    |   -D/--pgdata will be ignored if a repmgr configuration file is provided
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | NOTICE: destination directory "/bitnami/postgresql/data" provided
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | NOTICE: checking for available walsenders on the source node (2 required)
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | NOTICE: checking replication connections can be made to the source server (2 required)
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | WARNING: data checksums are not enabled and "wal_log_hints" is "off"
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | DETAIL: pg_rewind requires "wal_log_hints" to be enabled
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | NOTICE: starting backup (using pg_basebackup)...
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | pg_basebackup: error: directory "/data/oapi/ts_oapi" exists but is not empty
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | pg_basebackup: removing contents of data directory "/bitnami/postgresql/data"
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | NOTICE: replication slot "repmgr_slot_1001" deleted on node -1
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | ERROR: unable to take a base backup of the source server
pg-test_pg-1.1.aqsq6efqjcvi@docker03    | HINT: data directory ("/bitnami/postgresql/data") may need to be cleaned up manually

Even if I empty the /data/oapi/ts_oapi and /bitnami/postgresql/data the standby backend won't start. 

Seems that repmgr in this image is not able to work with external tablespaces? Or what I'm missing?
